### DISCUSSION TITLE
A New and Improved method of Handling Global menus.

### DISCUSSION THREAD

**Discussion by Weather-OS** [Weather-OS/Fildem-v2#5] [Ideas]
_View the issue at:_ https://github.com/Weather-OS/Fildem-v2/issues/3

## Create a new method for parsing the application's menu bar:

- [x] Recursively list all options on the menu bar, and store it within an accessible value. ([Implemented in a sub-merge](https://github.com/Weather-OS/Fildem-v2/commit/5079453e99a2824b25d7e134bba41e3840c9dc21))
- [x] Recursively parse options within the menu bar options and store them separately within a value. (https://github.com/Weather-OS/Fildem-v2/commit/945287eac21000ff5c4ec88fc018222539b62c05)
- [x] Create functional callers and events for the options to sync with the application. (https://github.com/Weather-OS/Fildem-v2/commit/945287eac21000ff5c4ec88fc018222539b62c05)

**Dbus**
- [ ] Add Dbus Signals to each member of `top_level_options`
- [ ] Receive Dbus calls from `extension.js`

**Top Level Options**
- [x] List Application's sub options. (Pre-implemented in `top_level_options`)
- [ ] List Switches, buttons and text boxes. 

## Designing the menu bar
- [ ] Follow GNOME's User shell theme
- [ ] List and place the menu bar options within the panel
- [ ] Add customizable margin support.
- [ ] Add an Auto hide feature.

## Bug-proof features
- [ ] Make sure that Fildem-V2 doesn't break other GNOME Extensions.
- [ ] Properly handle Fildem-V2's crashes.

### Optional macOS-Like features
- [ ] Add default menu members such as `File`, `Window`, `Help` to every spawned window (Might require patching).
- [ ] Add default menu members to GNOME Desktop.

**Comment by Weather-OS** [comment]
Recursively listing all options within a Window's Menu bar -> Complete.

Value: global.WindowMenuOptions;

Further information: It works perfectly fine

STDOUT: `[FILDEM_MENU] FILDEM: Top level Menus: ["_File","_Edit","_Selection","_View","_Go","_Run","_Terminal","_Help"]`

Successfully merged with the master branch.

**Reply by Weather-OS** [reply]
Recursively parsing options within the menu bar options and storing them separately within a value -> Complete.
Creating functional callers and events for the options to sync with the application -> Complete.

Value: `top_level_options` -> `_onSendTopLevelOptions(<dbus.ProxyObject> Proxy, <dbus.String> nameOwner, args)`

Accessible within: `menu_model.py` `extension.js`

Further information: Passes through perfectly fine.

STDOUT: `['_File  »  _New Text File', '_File  »  New File...', '_File  »  New _Window', '_File  »  ', '_File  »  _Open File...', '_File  »  Open _Folder... `...

Successfully merged with the master branch.

**Comment by Weather-OS** [comment]
Add Dbus Signals to each member of -> Almost done

Value: <dbus.ProxyObject.Signal> Activate(string _Option, <dbus.Interface> _Interface);

Accessible within: menu_model.py

Further information: Just needs some ground works.

STDOUT: No STDOUT Provided.

**Comment by ghost** [comment]
May I suggest adding native support for Firefox, Thunderbird and Zotero style menus? Those three apps share codebase from what I know. And they don't export the menus without some patching. Maybe it's possible to implement this from Fildem's side?
